keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots/README.md

```diff
@@ -0,0 +1,1132 @@
+      name: foo-testing
+      namespace: test
+    spec:
+      accessModes:
+      - ReadWriteOnce
+      resources:
+        requests:
+          storage: 10Mi
+      dataSourceRef:
+        apiGroup: snapshot.storage.k8s.io/v1alpha1
+        kind: VolumeSnapshotLink
+        name: foo-link
+      volumeMode: Filesystem
+    ```
+4. Once the populator finds a `VolumeSnapshotLink` is specified as `dataSourceRef`, it checks all `ReferenceGrants` in `VolumeSnapshotLink.spec.source.namespace` to see if populating the `VolumeSnapshotLink.spec.source` is allowed. If it is allowed, the populator populates the volume.
+    Note that how `ReferenceGrant` is checked depends on the implementation, however controllers that are trying to use the `VolumeSnapshot` in another namespace must check `ReferenceGrant` if the access is allowed before it actually starts exposing the data and metadata from the `VolumeSnapshot` to the `VolumeSnapshotLink`'s namespace.
+
+### API
+
+A new `VolumeSnapshotLink` CRD is introduced in `snapshot.storage.k8s.io` API group:
+
+```golang
+type VolumeSnapshotLink struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec VolumeSnapshotLinkSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+type VolumeSnapshotLinkList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items []VolumeSnapshotLink `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// VolumeSnapshotLinkSpec describes attributes of a volume snapshot link.
+type VolumeSnapshotLinkSpec struct {
+	// This field is immutable after creation.
+	Source VolumeSnapshotLinkSource `json:"source" protobuf:"bytes,1,opt,name=source"`
+}
+
+// VolumeSnapshotLinkSource specifies a reference to VolumeSnapshot.
+type VolumeSnapshotLinkSource struct {
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// Namespace is the namespace of the snapshot. When unspecified, the local namespace is inferred.
+	// Note that when a namespace is specified, a ReferenceGrant object is required in the referent namespace to allow that namespace's owner to accept the reference. See the ReferenceGrant documentation for details.
+	// +optional
+	Namespace string `json:"namespace" protobuf:"bytes,2,opt,name=namespace"`
+}
+```
+
+Note that when a `VolumeSnapshotLink.spec.source.namespace` is specified, a `ReferenceGrant` object is required in the referent namespace to allow that namespace’s owner to accept the reference. See the `ReferenceGrant` documentation for details.
+
+### Populator implementation
+
+The populator logic can be implemented either [(a) inside the existing CSI external-provisioner](#a-inside-the-existing-csi-external-provisioner) or [(b) as a separate populator](#b-as-a-separate-populator).
+Cluster admins can choose which implementation to be used per CSI driver basis.
```

Comment:
@msau42 
Is [this csi-provisioner code]([https://github.com/kubernetes-csi/external-](https://github.com/kubernetes-csi/external-provisioner/blob/master/cmd/csi-provisioner/csi-provisioner.go#L138)[provisioner/blob/master/cmd/csi-provisioner/csi-provisioner.go#L138](https://github.com/kubernetes-csi/external-provisioner/blob/master/cmd/csi-provisioner/csi-provisioner.go#L138)) getting feature gates from the cluster where it runs? And are you suggesting provisioner approach to also use the same mechanism to enable/disable this feature? If both answers are yes, populator approach can't provide a similar way to enable/disable this feature per cluster basis, because it relies on whether populator for the driver exists.

Reply:
Updated to use CrossNamespaceSourceProvisioning feature gate.